### PR TITLE
NATS: Acquire is too complicated

### PR DESCRIPTION
**Pull request by reinkrul**
Fixes https://github.com/nuts-foundation/nuts-node/issues/1615

Based on https://github.com/nuts-foundation/nuts-node/pull/1609

Some of the weirdness is gone now (recursive call) and time-outs are now properly supported, but not sure whether it's much better. Thoughts?

### CODE REVIEW

File: docs/pages/deployment/server_options.rst [hunk 1]
```diff
@@ -35,15 +35,15 @@
...
-    events.nats.timeout                    30                                                                                                                                                                                                                                                                                                                   Timeout for NATS server operations                                                                                                                                                                                                      
+    events.nats.timeout                    30ns                                                                                                                                                                                                                                                                                                                 Timeout, formatted as Golang duration (e.g. 5s, 1m), for NATS server operations
```
Comment (pascaldekloe) [inline docs/pages/deployment/server_options.rst:38]: Go, niet golang. Denk niet dat dit formaat iemand iets zegt. 😁 Beter de omschrijving van time.ParseDuration copieren hier.

File: events/conn.go [hunk 2]
```diff
@@ -81,36 +82,39 @@ func NewNATSConnectionPool(config Config) *NATSConnectionPool {
...
-			return pool.Acquire(ctx)
-		}
+	// use channels to synchronize connection attempts and allow timing out/cancellation while waiting.
+	ctx, cancel := context.WithTimeout(ctx, pool.config.Nats.Timeout)
+	defer cancel()
+	// either get the lock or time-out/cancelled
+	select {
+	case <-ctx.Done():
+		// Cancelled or time-out
+		return nil, nil, fmt.Errorf("time-out/cancelled while acquiring NATS connection: %w", ctx.Err())
```
Comment (pascaldekloe) [inline events/conn.go:97]: "NATS connection acquire: %w"

File: events/conn.go [hunk 3]
```diff
@@ -81,36 +82,39 @@ func NewNATSConnectionPool(config Config) *NATSConnectionPool {
...
-	log.Logger().Tracef("ConnectionState to %s", addr)
-
+	log.Logger().Tracef("Connecting to %s", addr)
 	for {
-		conn, err := pool.connectFunc(
-			addr,
-			nats.RetryOnFailedConnect(true),
-			nats.Timeout(time.Second*time.Duration(pool.config.Nats.Timeout)),
-		)
+		if ctx.Err() != nil {
```
Comment (pascaldekloe) [inline events/conn.go:113]: Check hardly applicable. Instead use a nats.Context in connectFunc instead.

File: events/conn.go [hunk 4]
```diff
@@ -81,36 +82,39 @@ func NewNATSConnectionPool(config Config) *NATSConnectionPool {
...
 	log.Logger().Trace("Trying to acquire a NATS connection")
-
 	// If the connection is already set, return it
-	data := pool.conn.Load()
-	if data != nil {
-		if conn, ok := data.(connectionAndContext); ok {
-			return conn.conn, conn.js, nil
-		}
+	if data := pool.conn.Load(); data != nil {
+		return data.(connectionAndContext).conn, data.(connectionAndContext).js, nil
```
Comment (pascaldekloe) [inline events/conn.go:87]: We have generics now. No need for panic.

**Review by pascaldekloe — CHANGES_REQUESTED** [review]

**Comment by reinkrul** [comment]
Not really worth the effort

**Pull request closed without merging**